File: algorithms/remove_element.py
#!/usr/bin/env python
# encoding: utf8

# Remove Element
# Given an array and a value, remove all instances of that value in place and return the new length.
# The order of elements can be changed. It doesn't matter what you leave beyond the new length.

__author__ = "Zhu Xianfeng <xianfeng.zhu@gmail.com>"

class Solution:
    # @param    A       a list of integers
    # @param    elem    an integer, value need to be removed
    # @return an integer
    def removeElement(self, A, elem):
        count = 0
        last  = -1
        idx   = -1
        for val in A:
            idx += 1
            if val == elem:
                if last == -1:
                    last = idx
                count += 1
            elif last != -1:
                # val not equals elem && last index is not -1
                A[last] = val
                last += 1
        return (idx + 1) - count

def main():
    num = [1, 3, 2, 3, 9, 2, 10]
    sol = Solution()
    count = sol.removeElement(num, 2)
    print num
    print count

if __name__ == "__main__":
    main()

File: algorithms/remove_element.cpp
/*
 * =====================================================================================
 *
 *       Filename:  remove_element.cpp
 *
 *    Description:  27. Remove Element. https://leetcode.com/problems/remove-element/
 *
 *        Version:  1.0
 *        Created:  02/02/2023 14:13:25
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  xianfeng.zhu@gmail.com
 *   Organization:
 *
 * =====================================================================================
 */
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"

using std::vector;

// Two pointers
class Solution {
 public:
  int removeElement(vector<int>& nums, int val) {
    int i = 0;
    for (int j = 0; j < nums.size(); j++) {
      if (nums[j] != val) {
        if (i != j) {
          nums[i] = nums[j];
        }
        i++;
      }
    }
    return i;
  }
};

TEST(Solution, removeElement) {
  vector<int> nums = {3, 2, 2, 3};
  int val = 3;
  const int k = Solution().removeElement(nums, val);
  EXPECT_EQ(k, 2);
}
